test/gtest/ucp/test_ucp_reconfigure.cc

```diff
@@ -0,0 +1,264 @@
+    }
+
+    bool is_exclude_iface() const
+    {
+        return get_variant_value();
+    }
+
+    void send_recv()
+    {
+        static constexpr unsigned num_iterations = 100;
+        static constexpr size_t msg_size         = 16 * UCS_KBYTE;
+        const ucp_request_param_t param          = {
+            .op_attr_mask = UCP_OP_ATTR_FLAG_NO_IMM_CMPL
+        };
+
+        for (unsigned i = 0; i < num_iterations; ++i) {
+            std::string sbuf(msg_size, 'a'), rbuf(msg_size, 'b');
+
+            void *sreq = ucp_tag_send_nbx(m_sender.ep(), sbuf.c_str(), msg_size,
+                                          0, &param);
+            void *rreq = ucp_tag_recv_nbx(m_receiver.worker(),
+                                          (void*)rbuf.c_str(), msg_size, 0, 0,
+                                          &param);
+            request_wait(rreq);
+            request_wait(sreq);
+            EXPECT_EQ(sbuf, rbuf);
+        }
+    }
+};
+
+void test_ucp_reconfigure::entity::store_config()
+{
+    for (ucp_lane_index_t lane = 0; lane < ucp_ep_num_lanes(ep()); ++lane) {
+        uct_ep_h uct_ep = ucp_ep_get_lane(ep(), lane);
+
+        /* Store UCT endpoints in order to compare with next configuration */
+        if (ucp_wireup_ep_test(uct_ep)) {
+            m_uct_eps.push_back(ucp_wireup_ep(uct_ep)->super.uct_ep);
+        } else {
+            m_uct_eps.push_back(uct_ep);
+        }
+    }
+
+    m_cfg_index = ep()->cfg_index;
+}
+
+ucp_tl_bitmap_t test_ucp_reconfigure::entity::ep_tl_bitmap() const
+{
+    ucp_tl_bitmap_t tl_bitmap = UCS_STATIC_BITMAP_ZERO_INITIALIZER;
+
+    if (ep() == NULL) {
+        return tl_bitmap;
+    }
+
+    for (ucp_lane_index_t lane = 0; lane < ucp_ep_num_lanes(ep()); ++lane) {
+        UCS_STATIC_BITMAP_SET(&tl_bitmap, ucp_ep_get_rsc_index(ep(), lane));
+    }
+
+    return tl_bitmap;
+}
+
+void test_ucp_reconfigure::entity::connect(const entity &other,
+                                           bool is_exclude_iface)
+{
+    auto worker_addr = other.get_address(false);
+    ucp_tl_bitmap_t tl_bitmap;
+    unsigned addr_indices[UCP_MAX_LANES];
+
+    tl_bitmap = is_exclude_iface ? UCS_STATIC_BITMAP_NOT(other.ep_tl_bitmap()) :
+                                   ucp_tl_bitmap_max;
+
+    UCS_ASYNC_BLOCK(&worker()->async);
+    ASSERT_UCS_OK(ucp_ep_create_to_worker_addr(worker(), &tl_bitmap,
+                                               &worker_addr->second,
+                                               UCP_EP_INIT_CREATE_AM_LANE,
+                                               "reconfigure test", addr_indices,
+                                               &m_ep));
+
+    m_ep->conn_sn = 0;
+    ASSERT_TRUE(ucp_ep_match_insert(worker(), m_ep, worker_addr->second.uuid,
+                                    m_ep->conn_sn, UCS_CONN_MATCH_QUEUE_EXP));
+
+    ASSERT_UCS_OK(ucp_wireup_send_request(m_ep));
+    UCS_ASYNC_UNBLOCK(&worker()->async);
+
+    store_config();

```

Comment:
should not store be under lock? AFAIU ep can be asynchronously reconfigured once wireup started